File: prismpath/kernel/level_m.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Level M, the match action fragment (SPEC section 4.3 and section 7): the classifier that says
whether a deterministic `when` condition is table encodable, the per edge report, and the per flow
verdict. This is the definition every compiler and interpreter target follows (prismpath-hw's
ppt_compile imports the chain desugar from here through model_check), so it sits below the analyzer
and the model checker and imports only the parser and the predicates."""
from __future__ import annotations

import ast
from typing import List, Optional, Tuple

from prismpath.kernel import predicates
from prismpath.kernel.parser import reachable as _reachable
from prismpath.kernel.predicates import expr_ast as _parse


# reason codes for non-membership (stable, machine-readable)
_R_CHAINED = "chained-comparison"
_R_FIELD_VS_FIELD = "field-vs-field"
_R_SUBSTRING = "substring-in"
_R_NONLITERAL = "non-literal-collection"
_R_STRING_ORDER = "string-ordering"
_R_CONSTANT = "constant-only"
_R_NESTED = "nested-container"
_R_SYNTAX = "disallowed-or-unparseable"

_ORDER_OPS = (ast.Lt, ast.LtE, ast.Gt, ast.GtE)
_EQ_OPS = (ast.Eq, ast.NotEq)


def _scalar_const(node) -> bool:
    # float is EXCLUDED: the Level M / hardware match-action fragment is an i32 value domain, so a float
    # constant is not table-representable (the PPT compiler delegates here and so rejects it too, as
    # `not-level-m:disallowed-or-unparseable`). bool is a subclass of int, so `isinstance(True, int)` is
    # True — booleans stay in-fragment.
    return isinstance(node, ast.Constant) and (
        node.value is None or (isinstance(node.value, (bool, int, str))
                               and not isinstance(node.value, float)))


def _atom_reason(node) -> Optional[str]:
    """None if `node` is a Level M atom; else the reason code it is not."""
    if isinstance(node, ast.Name):
        return None                                        # bare field (scalar truthiness)
    if isinstance(node, ast.Constant):
        return _R_CONSTANT                                 # `when True`: no field, not a row
    if isinstance(node, ast.Compare):
        if len(node.ops) != 1:
            return _R_CHAINED                              # defensive: is_level_m desugars chains
            #                                                first (SPEC §4.3), so this is unreachable
            #                                                on that path - kept so a raw _classify
            #                                                call can't silently read only ops[0].
        left, op, right = node.left, node.ops[0], node.comparators[0]
        if not isinstance(op, _ORDER_OPS + _EQ_OPS + (ast.In, ast.NotIn)):
            return _R_SYNTAX                               # `is`/`is not`: eval rejects them too
        # membership: field in/not in [scalar literals]
        if isinstance(op, (ast.In, ast.NotIn)):
            if not isinstance(left, ast.Name):
                return _R_FIELD_VS_FIELD if not isinstance(left, ast.Constant) else _R_CONSTANT
            if isinstance(right, ast.Constant) and isinstance(right.value, str):
                return _R_SUBSTRING                        # string RHS = substring test
            if isinstance(right, (ast.List, ast.Tuple)):
                for elt in right.elts:
                    if isinstance(elt, (ast.List, ast.Tuple)):
                        return _R_NESTED
                    if not _scalar_const(elt):
                        return _R_NONLITERAL
                return None
            return _R_NONLITERAL                           # membership in a runtime collection
        # comparison: field OP constant (either orientation)
        if isinstance(left, ast.Name) and _scalar_const(right):
            var_c = right
        elif isinstance(right, ast.Name) and _scalar_const(left):
            var_c = left
        elif isinstance(left, ast.Name) and isinstance(right, ast.Name):
            return _R_FIELD_VS_FIELD
        else:
            return _R_CONSTANT if (_scalar_const(left) and _scalar_const(right)) else _R_SYNTAX
        if isinstance(op, _ORDER_OPS) and isinstance(var_c.value, str):
            return _R_STRING_ORDER                         # string *ordering* is excluded
        return None
    return _R_SYNTAX


def _desugar_chains(node):
    """`a < b < c` -> `a < b and b < c`, recursively (SPEC §4.3: tooling SHOULD desugar chained
    comparisons before classifying/compiling). Exact under the engine's semantics: operands are pure
    (names/constants, evaluated identically each time), every pairwise comparison is total, and BoolOp
    evaluates all operands - so the desugared form cannot diverge from Python's chain evaluation on any
    context. This is the ONE desugar both the classifier (`is_level_m`) and the PPT compiler
    (`prismpath-hw/ppt_compile`, which imports this) use, so they can never disagree about chains."""
    if isinstance(node, ast.BoolOp):
        return ast.BoolOp(op=node.op, values=[_desugar_chains(val) for val in node.values])
    if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.Not):
        return ast.UnaryOp(op=node.op, operand=_desugar_chains(node.operand))
    if isinstance(node, ast.Compare) and len(node.ops) > 1:
        operands = [node.left] + list(node.comparators)
        return ast.BoolOp(op=ast.And(), values=[
            ast.Compare(left=operands[index], ops=[node.ops[index]], comparators=[operands[index + 1]])
            for index in range(len(node.ops))])
    return node


def _classify(node) -> Optional[str]:
    """None if the whole expression is a boolean combination of Level M atoms; else a reason."""
    if isinstance(node, ast.BoolOp):
        for val in node.values:
            reason = _classify(val)
            if reason is not None:
                return reason
        return None
    if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.Not):
        return _classify(node.operand)
    return _atom_reason(node)


def is_level_m(cond: str) -> Tuple[bool, Optional[str]]:
    """Is a deterministic condition in the match-action fragment? Keyword catch-alls
    (`always`/`else`/`false`…) are trivially table-encodable (the default / disabled row) and
    count as in-fragment. Non-deterministic conditions are (False, tier-name)."""
    if not predicates.is_deterministic(cond):
        if predicates.is_error(cond):
            expr = predicates.error_expr(cond)             # '' | 'when <expr>'
            if not expr:
                return True, None                          # bare `on error`: a table default row
            if not expr.lower().startswith("when "):
                expr = "when " + expr
            return is_level_m(expr)
        return False, "not-deterministic"
    expr = predicates._expr_of(cond)
    if expr.lower() in predicates.ALWAYS or expr.lower() in predicates.NEVER:
        return True, None
    node = _parse(cond)
    if node is None:
        return False, _R_SYNTAX
    node = _desugar_chains(node)             # SPEC §4.3: chained comparisons normalize into the fragment
    reason = _classify(node)
    return (reason is None), reason


def level_m_report(graph) -> List[dict]:
    """Per-edge fragment membership for every reachable deterministic edge (SPEC §4.3: tooling
    SHOULD report this). Rows: {node, target, condition, level_m, reason}."""
    out = []
    reach = _reachable(graph)
    for name in sorted(reach):
        node = graph.nodes.get(name)
        if node is None:
            continue
        for target, condition in node.edges:
            if not predicates.is_deterministic(condition):
                continue
            ok, reason = is_level_m(condition)
            out.append({"node": name, "target": target, "condition": condition,
                        "level_m": ok, "reason": reason})
    return out


def flow_level_m(graph) -> Tuple[bool, List[dict]]:
    """SPEC §7: within P0, Level M marks flows whose deterministic edges are ALL in the
    fragment. Returns (all_in_fragment, non-member rows)."""
    rows = level_m_report(graph)
    bad = [row for row in rows if not row["level_m"]]
    return (not bad), bad

File: prismpath-rs/src/level_m.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Level M proof layer and capability reporting.

use crate::condition::{error_expr, expr_of, is_deterministic, is_error, parse_expr, Ast};
use crate::engine::portability_violations;
use crate::parser::Graph;
use crate::value::Value;

mod lm_reason {
    pub const CHAINED: &str = "chained-comparison";
    pub const FIELD_VS_FIELD: &str = "field-vs-field";
    pub const SUBSTRING: &str = "substring-in";
    pub const NONLITERAL: &str = "non-literal-collection";
    pub const STRING_ORDER: &str = "string-ordering";
    pub const CONSTANT: &str = "constant-only";
    pub const NESTED: &str = "nested-container";
    pub const SYNTAX: &str = "disallowed-or-unparseable";
}

fn lm_is_order_op(op: &str) -> bool {
    matches!(op, "<" | "<=" | ">" | ">=")
}

fn lm_scalar_const(node: &Ast) -> bool {
    matches!(
        node,
        Ast::Const(Value::Null, _)
            | Ast::Const(Value::Bool(_), _)
            | Ast::Const(Value::Str(_), _)
            | Ast::Const(Value::Num(_), false)
    )
}

fn lm_desugar_chains(node: Ast) -> Ast {
    match node {
        Ast::And(vals) => Ast::And(vals.into_iter().map(lm_desugar_chains).collect()),
        Ast::Or(vals) => Ast::Or(vals.into_iter().map(lm_desugar_chains).collect()),
        Ast::Not(val) => Ast::Not(Box::new(lm_desugar_chains(*val))),
        Ast::Cmp { left, ops, rights } if ops.len() > 1 => {
            let operands: Vec<Ast> = std::iter::once(*left).chain(rights).collect();
            let vals = ops
                .iter()
                .enumerate()
                .map(|(idx, op)| Ast::Cmp {
                    left: Box::new(operands[idx].clone()),
                    ops: vec![op.clone()],
                    rights: vec![operands[idx + 1].clone()],
                })
                .collect();
            Ast::And(vals)
        }
        other => other,
    }
}

fn lm_atom_reason(node: &Ast) -> Option<&'static str> {
    match node {
        Ast::Name(_) => None,
        Ast::Const(..) => Some(lm_reason::CONSTANT),
        Ast::Cmp { left, ops, rights } => {
            if ops.len() != 1 {
                return Some(lm_reason::CHAINED);
            }
            let op = ops[0].as_str();
            let right = &rights[0];
            if op == "in" || op == "not in" {
                match left.as_ref() {
                    Ast::Name(_) => {}
                    Ast::Const(..) => return Some(lm_reason::CONSTANT),
                    _ => return Some(lm_reason::FIELD_VS_FIELD),
                }
                if let Ast::Const(Value::Str(_), _) = right {
                    return Some(lm_reason::SUBSTRING);
                }
                if let Ast::List(elts) = right {
                    for elem in elts {
                        if let Ast::List(_) = elem {
                            return Some(lm_reason::NESTED);
                        }
                        if !lm_scalar_const(elem) {
                            return Some(lm_reason::NONLITERAL);
                        }
                    }
                    return None;
                }
                return Some(lm_reason::NONLITERAL);
            }
            let left_is_name = matches!(left.as_ref(), Ast::Name(_));
            let right_is_name = matches!(right, Ast::Name(_));
            let var_const: &Ast = if left_is_name && lm_scalar_const(right) {
                right
            } else if right_is_name && lm_scalar_const(left) {
                left
            } else if left_is_name && right_is_name {
                return Some(lm_reason::FIELD_VS_FIELD);
            } else {
                return Some(if lm_scalar_const(left) && lm_scalar_const(right) {
                    lm_reason::CONSTANT
                } else {
                    lm_reason::SYNTAX
                });
            };
            if lm_is_order_op(op) {
                if let Ast::Const(Value::Str(_), _) = var_const {
                    return Some(lm_reason::STRING_ORDER);
                }
            }
            None
        }
        _ => Some(lm_reason::SYNTAX),
    }
}

fn lm_classify(node: &Ast) -> Option<&'static str> {
    match node {
        Ast::And(vals) | Ast::Or(vals) => {
            for item in vals {
                if let Some(reason_str) = lm_classify(item) {
                    return Some(reason_str);
                }
            }
            None
        }
        Ast::Not(val) => lm_classify(val),
        _ => lm_atom_reason(node),
    }
}

pub fn is_level_m(cond: &str) -> (bool, Option<String>) {
    if !is_deterministic(cond) {
        if is_error(cond) {
            let mut expr = error_expr(cond);
            if expr.is_empty() {
                return (true, None);
            }
            if !expr.to_lowercase().starts_with("when ") {
                expr = format!("when {expr}");
            }
            return is_level_m(&expr);
        }
        return (false, Some("not-deterministic".to_string()));
    }
    let expr = expr_of(cond);
    let low = expr.to_lowercase();
    if ["always", "true", "else", "otherwise", "default", "_"].contains(&low.as_str())
        || ["false", "never"].contains(&low.as_str())
    {
        return (true, None);
    }
    let node = match parse_expr(&expr) {
        Ok(n_ast) => lm_desugar_chains(n_ast),
        Err(_) => return (false, Some(lm_reason::SYNTAX.to_string())),
    };
    match lm_classify(&node) {
        None => (true, None),
        Some(reason_str) => (false, Some(reason_str.to_string())),
    }
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct NonMemberEdge {
    pub node: String,
    pub target: String,
    pub condition: String,
    pub level_m: bool,
    pub reason: Option<String>,
}

pub fn flow_level_m(graph: &Graph) -> (bool, Vec<NonMemberEdge>) {
    let mut bad = Vec::new();
    let mut names = crate::engine::reachable(graph);
    names.sort();
    for name in names {
        let Some(node) = graph.nodes.get(&name) else { continue };
        for (target, cond) in &node.edges {
            if !is_deterministic(cond) {
                continue;
            }
            let (level_m_ok, reason) = is_level_m(cond);
            if !level_m_ok {
                bad.push(NonMemberEdge {
                    node: name.clone(),
                    target: target.clone(),
                    condition: cond.clone(),
                    level_m: false,
                    reason,
                });
            }
        }
    }
    (bad.is_empty(), bad)
}

pub fn capability_report(graph: &Graph) -> serde_json::Value {
    let semantic = portability_violations(graph);
    let p0 = semantic.is_empty();
    let (lm_ok, non_member) = flow_level_m(graph);
    let hw_ok = p0 && lm_ok;

    let semantic_json: Vec<serde_json::Value> = semantic
        .iter()
        .map(|violation| serde_json::json!({ "node": violation.node, "target": violation.target, "condition": violation.condition }))
        .collect();
    let non_member_json =
        serde_json::to_value(&non_member).unwrap_or(serde_json::Value::Array(vec![]));

    let portable_reason = if p0 {
        serde_json::Value::Null
    } else {
        serde_json::json!(format!(
            "{} reachable semantic edge(s) \u{2014} P0 runs unconditionally; lock them for P1",
            semantic.len()
        ))
    };
    let hw_reason = if hw_ok {
        serde_json::Value::Null
    } else if !p0 {
        serde_json::json!(format!(
            "{} reachable semantic edge(s) \u{2014} not deterministic",
            semantic.len()
        ))
    } else {
        serde_json::json!(format!(
            "{} deterministic edge(s) outside the match-action fragment",
            non_member.len()
        ))
    };

    serde_json::json!({
        "tier": if p0 { "P0" } else { "P1/P2" },
        "level_m": lm_ok,
        "targets": {
            "python": { "status": "yes", "reason": null, "blocking_edges": [] },
            "portable": {
                "status": if p0 { "yes" } else { "needs-lockfile" },
                "reason": portable_reason,
                "blocking_edges": if p0 { serde_json::json!([]) } else { serde_json::json!(semantic_json) },
            },
            "level_m_hardware": {
                "status": if hw_ok { "yes" } else { "no" },
                "reason": hw_reason,
                "blocking_edges": if !p0 {
                    serde_json::json!(semantic_json)
                } else if hw_ok {
                    serde_json::json!([])
                } else {
                    non_member_json
                },
            },
        },
    })
}
